jbmc/src/java_bytecode/assignments_from_json.cpp

```diff
@@ -69,65 +69,55 @@ followed_class_type(const exprt &expr, const symbol_table_baset &symbol_table)
   return java_class_type;
 }
 
-static bool
-has_array_type(const exprt &expr, const symbol_table_baset &symbol_table)
-{
-  return has_prefix(
-    id2string(followed_class_type(expr, symbol_table).get_tag()),
-    "java::array[");
-}
-
-static bool is_enum_type(const java_class_typet &class_type)
-{
-  return class_type.get_base("java::java.lang.Enum").has_value();
-}
-
 static bool
 has_enum_type(const exprt &expr, const symbol_table_baset &symbol_table)
 {
-  return is_enum_type(followed_class_type(expr, symbol_table));
+  return followed_class_type(expr, symbol_table).get_is_enumeration();
 }
 
 /// This function is used as a workaround until reference-equal objects defined
 /// across several classes are tracked correctly. Once reference-equality works
 /// in all cases, this function can be removed.
 /// Until then, in the case of an enum expression that needs to be assigned a
 /// value, we distinguish between two cases:
-/// 1) the enum expression is declared in a class of its own type - in this
-///    case, initialize it just as a regular object that has known reference-
-///    equal objects. (Corresponds to creating the enum constant in Java.)
-///    See \ref assign_reference_from_json.
+/// 1) the enum expression is declared in its own type - in this case,
+///    initialize it just as a regular object that has known reference-equal
+///    objects. (Corresponds to creating the enum constant in Java or
+///    referencing it directly.)
+///    See assign_reference_from_json.
 /// 2) otherwise, initialize it by indexing the $VALUES array with the given
-///    ordinal. (Corresponds to retrieving the enum constant in Java.)
-///    See \ref assign_enum_from_json.
+///    ordinal.
+///    See assign_enum_from_json.
 /// \param expr: an expression representing a Java object.
 /// \param symbol_table: used for looking up the type of \p expr.
 /// \param declaring_class_type: type of the class where \p expr is declared.
 /// \return `true` if \p expr has an enum type and is declared within the
 ///   definition of that same type, `false` otherwise.
-bool is_enum_definition(
+bool is_enum_declared_in_own_type(
   const exprt &expr,
   const symbol_table_baset &symbol_table,
   const java_class_typet &declaring_class_type)
 {
   PRECONDITION(can_cast_type<pointer_typet>(expr.type()));
   return followed_class_type(expr, symbol_table) == declaring_class_type &&
-         is_enum_type(declaring_class_type);
+         declaring_class_type.get_is_enumeration();
 }
 
-/// Returns true iff the argument has a "@type" key.
+/// Return true iff the argument has a "@type" key.
```

Comment:
:no_entry_sign: Documentation needs updating. It's now an optional.

Reply:
Changed to
```
/// If the argument has a "@type" key, return the corresponding value, else
/// return an empty optional.
```